planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/configschema/TagCriteria.java

```diff
@@ -0,0 +1,59 @@
+package com.onthegomap.planetiler.custommap.configschema;
+
+import static com.onthegomap.planetiler.expression.Expression.matchAnyTyped;
+import static com.onthegomap.planetiler.expression.Expression.matchField;
+
+import com.onthegomap.planetiler.custommap.TagValueProducer;
+import com.onthegomap.planetiler.expression.Expression;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+public class TagCriteria {
+
+  /**
+   * Returns a function that determines whether a source feature matches any of the entries in this specification
+   * 
+   * @param map              a map of tag criteria
+   * @param tagValueProducer a TagValueProducer
+   * @return a predicate which returns true if this criteria matches
+   */
+  public static Expression matcher(Map<String, Object> map, TagValueProducer tagValueProducer) {
+
+    if (map.isEmpty()) {
+      return Expression.TRUE;
+    }
+
+    List<Expression> tagExpressions = new ArrayList<>();
+
+    map.entrySet()
```

Comment:
A standard `for (var entry : map.entrySet()) { ...` might be more straightforward here.

Reply:
@msbarry major refactor in 848507a, please review and see what you think?